### PR TITLE
8341445: DecimalFormatSymbols setters should throw NPE

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
Please review this PR which improves the safety of equality checking for DecimalFormatSymbols. As certain setters did not throw NPE, this allowed for NPE in the equality method. This PR now updates the setters to throw NPE.

In addition to the NPEs, there is also a behavioral change that `setInternationalCurrencySymbol` no longer sets currency to null if the `currencyCode` is invalid. Instead, it simply does not update `currency`. This must be done, because we do not want to allow nullable instance variables post `initalizeCurrency`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8341801](https://bugs.openjdk.org/browse/JDK-8341801) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8341445](https://bugs.openjdk.org/browse/JDK-8341445): DecimalFormatSymbols setters should throw NPE (**Bug** - P4)
 * [JDK-8341801](https://bugs.openjdk.org/browse/JDK-8341801): DecimalFormatSymbols setters should throw NPE (**CSR**)

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21315/head:pull/21315` \
`$ git checkout pull/21315`

Update a local copy of the PR: \
`$ git checkout pull/21315` \
`$ git pull https://git.openjdk.org/jdk.git pull/21315/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21315`

View PR using the GUI difftool: \
`$ git pr show -t 21315`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21315.diff">https://git.openjdk.org/jdk/pull/21315.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21315#issuecomment-2389968486)

### CODE REVIEW

**Comment by liach** [comment]
Should we add an explicit whitelist of methods allowed to throw NPE, so that if we add new getter/setters to DFS, we can avoid this nasty null problem for the future? This whitelist will be validated by tests for sure.

**Comment by justin-curtis-lu** [comment]
Thanks for taking a look @liach . https://github.com/openjdk/jdk/pull/21315/commits/bd6b021cc84034d55eb6d262ade7ea19e520c1fe adds a test that ensures if a new NPE throwing setter is added without being whitelisted, this test will now fail. This should prevent mishaps like this in the future. While this feels a little overkill, this enforcement is done via a test, so it is fine by me.

**Comment by naotoj** [comment]
I am not sure allowing null for those setters just because not to throw exceptions in equals() is the right solutoin. The distinction is made purely based on the existing behavior (which I believe is just an overlook in the first place). My preference is to null check in the setter methods, as there is no reason to set them null. That is just a misuse of the setters.

**Comment by justin-curtis-lu** [comment]
> I am not sure allowing null for those setters just because not to throw exceptions in equals() is the right solutoin. The distinction is made purely based on the existing behavior (which I believe is just an overlook in the first place). My preference is to null check in the setter methods, as there is no reason to set them null. That is just a misuse of the setters.

OK, if we are fine with the behavioral change, I will change the direction of this PR to instead be an update to the setters to ensure NPEs for the offending methods. Will file a CSR.

**Comment by justin-curtis-lu** [comment]
https://github.com/openjdk/jdk/pull/21315/commits/91b1c78f0cfe4fe866c5eca3e84bf42e4c45adf1 updates this PR to take the throws NPE for all setters route. Please see the CSR filed: https://bugs.openjdk.org/browse/JDK-8341801.

Also note that there is an additional behavioral change that `setInternationalCurrencySymbol(String currencyCode)` will now not update the currency field if currencyCode is invalid, rather than nulling the currency field. This must be done to ensure that currency is not null (post lazy loading of the currency fields).

File: src/java.base/share/classes/java/text/DecimalFormatSymbols.java [hunk 1]
```diff
@@ -767,7 +763,8 @@ public boolean equals(Object obj) {
             patternSeparator == other.patternSeparator &&
             infinity.equals(other.infinity) &&
             NaN.equals(other.NaN) &&
-            getCurrencySymbol().equals(other.getCurrencySymbol()) && // possible currency init occurs here
+            // Currency fields are lazy. Init via get call to ensure non-null
+            getCurrencySymbol().equals(other.getCurrencySymbol()) &&
```
Comment (naotoj) [inline src/java.base/share/classes/java/text/DecimalFormatSymbols.java:767]: Probably the same comment refinement can be applied to the location in `hashCode()`.

File: test/jdk/java/text/Format/DecimalFormat/SettersShouldThrowNPETest.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class SettersShouldThrowNPETest {
+
+    // The public setter methods that should throw NPE
+    private static final List<Method> NPE_SETTERS =
+            Arrays.stream(DecimalFormatSymbols.class.getDeclaredMethods())
+            .filter(m -> Modifier.isPublic(m.getModifiers()))
+            .filter(m -> m.getName().startsWith("set"))
+            .filter(m -> Stream.of(m.getParameterTypes()).noneMatch(Class::isPrimitive))
```
Comment (naotoj) [inline test/jdk/java/text/Format/DecimalFormat/SettersShouldThrowNPETest.java:55]: Can consolidate into a single filter()?

**Review by naotoj — COMMENTED** [review]
Looks good overall. The test can also check the behavioral change in `setInternationalCurrencySymbol()`.

File: src/java.base/share/classes/java/text/DecimalFormatSymbols.java (hunk 1 above)
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/text/DecimalFormatSymbols.java:767]: This and the other comments addressed via https://github.com/openjdk/jdk/pull/21315/commits/d31cb2ebc5eb92978bed4c6f38f1ddfd7ff5452c. Thanks!

**Review by naotoj — APPROVED** [review]

**Pull request closed without merging**